scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+{
+	struct task_ctx *tctx;
+
+	if ((tctx = bpf_task_storage_get(&task_ctxs, p, 0, 0))) {
+		return tctx;
+	}
+
+	scx_bpf_error("task_ctx lookup failed");
+	return NULL;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctxs SEC(".maps");
+
+static inline struct cpu_ctx *lookup_cpu_ctx(int cpu)
+{
+	struct cpu_ctx *cctx;
+	u32 zero = 0;
+
+	if (cpu < 0)
+		cctx = bpf_map_lookup_elem(&cpu_ctxs, &zero);
+	else
+		cctx = bpf_map_lookup_percpu_elem(&cpu_ctxs, &zero, cpu);
+
+	if (!cctx) {
+		scx_bpf_error("no cpu_ctx for cpu %d", cpu);
+		return NULL;
+	}
+
+	return cctx;
+}
+
+struct cell {
+	u64 vtime_now;
+	u32 dsq;
+	// The following field is populated from userspace to indicate
+	// which cpus the cell should belong to.
+	unsigned char cpus[MAX_CPUS_U8];
+};
+
+struct cell cells[MAX_CELLS];
+
+static inline struct cell *lookup_cell(int idx)
+{
+	if (idx < 0 || idx >= MAX_CELLS) {
+		scx_bpf_error("invalid cell %d", idx);
+		return NULL;
+	}
+	return &cells[idx];
+}
+
+/*
+ * Store the cpumask for each cell (owned by BPF logic)
+ */
+struct cell_cpumask_wrapper {
+	struct bpf_cpumask __kptr *cpumask;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct cell_cpumask_wrapper);
+	__uint(max_entries, MAX_CELLS);
+	__uint(map_flags, 0);
+} cell_cpumasks SEC(".maps");
+
+static inline struct cpumask *lookup_cell_cpumask(int idx)
+{
+	struct cell_cpumask_wrapper *cpumaskw;
+
+	if (!(cpumaskw = bpf_map_lookup_elem(&cell_cpumasks, &idx))) {
+		scx_bpf_error("no cell cpumask");
+		return NULL;
+	}
+
+	return (struct cpumask *)cpumaskw->cpumask;
+}
+
+/* css iteration is experimental, forward declare the necessary functions */
+struct bpf_iter_css;
+struct cgroup_subsys_state;
+extern int bpf_iter_css_new(struct bpf_iter_css *it,
+			    struct cgroup_subsys_state *start,
+			    unsigned int flags) __weak __ksym;
+extern struct cgroup_subsys_state *
+bpf_iter_css_next(struct bpf_iter_css *it) __weak __ksym;
+extern void bpf_iter_css_destroy(struct bpf_iter_css *it) __weak __ksym;
+
+/*
+* Along with a user_global_seq bump, indicates that cgroup->cell assignment
+* changed
+*/
+volatile bool update_cell_assignment;
+bool draining;
+
+/*
+ * This is the main driver for reconfiguration. It only runs on CPU 0
+ * To avoid
+ */
+SEC("fentry/scheduler_tick")
+int scheduler_tick_fentry(const void *ctx)
+{
+	int cell_idx, cpu_idx;
+	struct cpu_ctx *cpu_ctx;
+	struct bpf_cpumask *cpumask;
+	struct cell_cpumask_wrapper *cell_cpumaskw;
+	struct cgroup_subsys_state *root_css, *pos;
+	struct cgroup *root_cgrp;
+
+	if (bpf_get_smp_processor_id() != 0)
+		return 0;
+
+	/*
+	 * To handle races where tasks are assigned to cells that are getting
+	 * removed, we ensure cpus dispatch from their previous cell for an entire
+	 * scheduler tick. This is a crude way of mimicing RCU synchronization.
```

Comment:
Hmm, not sure I follow the suggestion. Just because I see no matching tasks in the DSQ doesn't mean that there can't be another task going through select_cpu having read the old cg -> cell assignment and about to dispatch to the dying cell, no?

Reply:
Yeah, that's the only real solution I see without heavier weight synchronization.